File: core-linux/crates/ss-core-ipc/src/dto.rs
//! DTO types for the IPC contract (mirrored on the Dart side via `json_serializable`).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CoreStatus {
    pub version: String,
    pub uptime_seconds: u64,
    pub tpm_used: bool,
    pub db_ok: bool,
    pub last_enforcement_error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChildStatus {
    pub today_minutes_used: u32,
    pub today_budget_minutes: Option<u32>,
    pub current_window_open: bool,
    pub current_window_ends_at: Option<DateTime<Utc>>,
    pub active_blocklist_display: Vec<String>,
    pub session_running: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DailyReport {
    pub date: chrono::NaiveDate,
    pub usage_by_app: Vec<AppUsage>,
    pub total_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppUsage {
    pub content_hash: String,
    pub display_name: Option<String>,
    pub basename: String,
    pub minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FamilySnapshot {
    pub family_id: Uuid,
    pub family_name: String,
    pub parents: Vec<ParentView>,
    pub children: Vec<ChildView>,
    pub devices: Vec<DeviceView>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParentView {
    pub id: Uuid,
    pub display_name: String,
    // auth_hash is NEVER included here (security rule §8.4).
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChildView {
    pub id: Uuid,
    pub display_name: String,
    pub birth_year: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceView {
    pub id: Uuid,
    pub child_id: Uuid,
    pub hostname: String,
    pub platform: String,
    pub last_seen_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OnboardingState {
    Required,
    Complete,
}

File: ui/lib/ipc/dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'dto.freezed.dart';
part 'dto.g.dart';

@freezed
abstract class CoreStatus with _$CoreStatus {
  const factory CoreStatus({
    required String version,
    @JsonKey(name: 'uptime_seconds') required int uptimeSeconds,
    @JsonKey(name: 'tpm_used') required bool tpmUsed,
    @JsonKey(name: 'db_ok') required bool dbOk,
    @JsonKey(name: 'last_enforcement_error') String? lastEnforcementError,
  }) = _CoreStatus;
  factory CoreStatus.fromJson(Map<String, dynamic> j) =>
      _$CoreStatusFromJson(j);
}

@freezed
abstract class ChildStatus with _$ChildStatus {
  const factory ChildStatus({
    @JsonKey(name: 'today_minutes_used') required int todayMinutesUsed,
    @JsonKey(name: 'today_budget_minutes') int? todayBudgetMinutes,
    @JsonKey(name: 'current_window_open') required bool currentWindowOpen,
    @JsonKey(name: 'current_window_ends_at') DateTime? currentWindowEndsAt,
    @JsonKey(name: 'active_blocklist_display')
    required List<String> activeBlocklistDisplay,
    @JsonKey(name: 'session_running') required bool sessionRunning,
  }) = _ChildStatus;
  factory ChildStatus.fromJson(Map<String, dynamic> j) =>
      _$ChildStatusFromJson(j);
}

@freezed
abstract class AppUsage with _$AppUsage {
  const factory AppUsage({
    @JsonKey(name: 'content_hash') required String contentHash,
    @JsonKey(name: 'display_name') String? displayName,
    required String basename,
    required int minutes,
  }) = _AppUsage;
  factory AppUsage.fromJson(Map<String, dynamic> j) => _$AppUsageFromJson(j);
}

@freezed
abstract class DailyReport with _$DailyReport {
  const factory DailyReport({
    required DateTime date,
    @JsonKey(name: 'usage_by_app') required List<AppUsage> usageByApp,
    @JsonKey(name: 'total_minutes') required int totalMinutes,
  }) = _DailyReport;
  factory DailyReport.fromJson(Map<String, dynamic> j) =>
      _$DailyReportFromJson(j);
}
